ga4gh/backends.py

```diff
@@ -309,6 +310,109 @@ def getMetadata(self):
         return ret
 
 
+def _encode_value(value):
+    if isinstance(value, (list, tuple)):
+        return [ str(v) for v in value ]
+    else:
+        return [str(value)]
+
+
+class HtslibVariantSet(VariantSet):
+    """
+    Class representing a single variant set backed by a tabix directory.
+    """
+    def __init__(self, variantSetId, vcfPath, use_bcf=True):
+        self._variantSetId = variantSetId
+        self._created = protocol.convertDatetime(datetime.datetime.now())
+        self._chromFileMap = {}
+
+        for pattern in ['*.bcf', '*.vcf.gz']:
+            for vcfFile in glob.glob(os.path.join(vcfPath, pattern)):
+                self._addFile(vcfFile)
+
+    def _addFile(self, vcfFile):
+        varFile = pysam.BCFFile(vcfFile)
+        if not varFile.index:
+            raise Exception("VCF/BCF files must be indexed")
+        for chrom in varFile.index:
+            has_records = bool(next(varFile.fetch(chrom), None))
```

Comment:
Yes; the Tabix index lists only contigs that are present in the indexed
file, while the CSI index (used by BCF) includes records for all contigs
declared in the BCF header.  Without such a check, adding BCF files will
likely fail due to spurious overlapping contig errors.  Likewise, this is
why the check is so application specific, since most applications don't
need to precompute the subset of contigs actually present in a BCF file.

On Wed, Jan 14, 2015 at 8:47 AM, Jerome Kelleher <notifications@github.com>
wrote:

> In ga4gh/backends.py
> <https://github.com/jeromekelleher/server/pull/1#discussion-diff-22949915>
> :
>
> > +    """
> > +    def __init__(self, variantSetId, vcfPath, use_bcf=True):
> > +        self._variantSetId = variantSetId
> > +        self._created = protocol.convertDatetime(datetime.datetime.now())
> > +        self._chromFileMap = {}
> > +
> > +        for pattern in ['*.bcf', '*.vcf.gz']:
> > +            for vcfFile in glob.glob(os.path.join(vcfPath, pattern)):
> > +                self._addFile(vcfFile)
> > +
> > +    def _addFile(self, vcfFile):
> > +        varFile = pysam.BCFFile(vcfFile)
> > +        if not varFile.index:
> > +            raise Exception("VCF/BCF files must be indexed")
> > +        for chrom in varFile.index:
> > +            has_records = bool(next(varFile.fetch(chrom), None))
>
> It's up to you @bioinformed <https://github.com/bioinformed>, whatever
> you think. Do we really need to defend against this case anyway?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jeromekelleher/server/pull/1/files#r22949915>.
>